### PR TITLE
Better bucket prefix support

### PR DESCRIPTION
**Pull request by netsettler**
This is a branch-to-branch PR to add better bucket prefix support for `s3_utils`, `deployment_utils, and (opportunistically), `beanstalk_utils`.

It might need more unit tests, of `s3_utils`, but hopefully for now we can substitute some real-world deploys.

These are the specific changes (copied from the `CHANGELOG.rst`...

* In `beanstalk_utils`:

  * (Opportunistic) Add `'elasticbeanstalk-%s-metadata-bundles'` to the list of buckets that `beanstalk_utils.delete_s3_buckets` is willing to delete.  (This functionality still only works for legacy systems so it may not matter, but if we're going to have code like this retained at all, it should be maintained.)

* In `cloudformation_utils`:

  * New functions `camelize` and `dehyphenate` because they're needed a lot in our `4dn-cloud-infra` repo.

  * New implementation of functions `get_ecs_real_url` and `get_ecr_repo_url` that are not Alpha-specific.

  * New classes `AbstractOrchestrationManager`, `C4OrchestrationManager`, and `AwsemOrchestrationManager` with various utilities ported from `4dn-cloud-infra` (so they could be used to re-implement `get_ecs_real_url`and `get_ecr_repo_url`).

  * New `test_cloudformation_utils.py` testing each of the bits of functionality in `cloudformation_utils` along normal paths, including sometimes mocking both the Alpha and KMP environments, hoping transitions will be smooth. (I removed `test_cfn_utils.py` which was confusingly named (I overlooked it when looking for tests) and didn't contain much anyway. What little it had was obsolete not because of functionality but the mechanism of mocking didn't survive my reimplementation, so I tried to make an equivalent test case in the new version.)

* In `deployment_utils`:

  * Support environment variable `ENCODED_TIBANNA_OUTPUT_BUCKET` and `--tibanna_output_bucket` to control environment variable `$TIBANNA_OUTPUT_BUCKET` in construction of `production.ini`.

  * Support environment variable `ENCODED_APPLICATION_BUCKET_PREFIX` and `--application_bucket_prefix` to control environment variable `$APPLICATION_BUCKET_PREFIX` in construction of `production.ini`.

  * Support environment variable `ENCODED_FOURSIGHT_BUCKET_PREFIX` and `--foursight_bucket_prefix` to control environment variable `$FOURSIGHT_BUCKET_PREFIX` in construction of `production.ini`.

  * New class variable `APP_KIND` in `IniFileManager`. Default is `None`, but new subclasses adjust the default to `cgap` or `fourfront`.

  * New class variable `APP_ORCHESTRATED` in `IniFileManager`. Default is `None`, but new subclasses adjust the default to `True` or `False`.

  * New classes

    * `BasicCGAPIniFileManager`
    * `BasicLegacyCGAPIniFileManager`
    * `BasicOrchestratedCGAPIniFileManager`
    * `BasicFourfrontIniFileManager`
    * `BasicLegacyFourfrontIniFileManager`
    * `BasicOrchestratedFourfrontIniFileManager`

    In principle, this should allow some better defaulting.

* In `qa_utils`, support for mocking enough of `boto3.client('cloudformation')` that we can test
  `cloudformation_utils`.

* In `s3_utils`:

  * Make initialize attribute `.metadata_bucket` better.

  * Add an attribute `.tibanna_output_bucket` and initialize it properly from the health page.

### CODE REVIEW

File: dcicutils/deployment_utils.py [hunk 1]
```diff
@@ -802,6 +855,30 @@ def main(cls):
...
+class BasicFourfrontIniFileManager(IniFileManager):
+    APP_KIND = 'fourfront'
+
+
+class BasicLegacyFourfrontIniFileManager(BasicFourfrontIniFileManager):
+    APP_ORCHESTRATED = False
+
+
+class BasicOrchestratedFourfrontIniFileManager(BasicFourfrontIniFileManager):
+    APP_ORCHESTRATED = True
```
Comment (willronchetti) [inline dcicutils/deployment_utils.py:879]: Although it's clear reading the code, I would comment all of these describing what it is. For example
```
class BasicOrchestratedCGAPIniFileManager(BasicCGAPIniFileManager):
""" production.ini builder for CGAP environments in the Alpha infrastructure (4dn-cloud-infra) """
    APP_ORCHESTRATED = True
```

**Review by willronchetti — APPROVED** [review]

**Comment by netsettler** [comment]
@willronchetti Sorry, didn't even notice my comments on this PR had not been pushed through, so you didn't have any of my guidance to help you read through it.

Also, I have added [one more commit](https://github.com/4dn-dcic/utils/pull/151/commits/e2223bd40bbe84e17e3bb9cd0564b73cb45ad577) that may need additional scrutiny.

File: dcicutils/deployment_utils.py (hunk 1 above)
Comment (netsettler) [reply dcicutils/deployment_utils.py]: I added doc strings.

File: pyproject.toml [hunk 2]
```diff
@@ -1,6 +1,6 @@
 [tool.poetry]
 name = "dcicutils"
-version = "1.20.0.1b8"
+version = "1.20.0.1b14"  # to become 1.21.0 when checked in
```
[the remark this answers is not in the document]
Comment (netsettler) [reply pyproject.toml]: We agreed we'd use 2.0

File: dcicutils/cloudformation_utils.py [hunk 3]
```diff
@@ -1,38 +1,223 @@
...
+    if proper conventions were followed in by tagging the environment's ECS stack with an output
+    generated by make_required_key_for_ecs_application_url.
+    """
+    manager = C4OrchestrationManager()
+    found = manager.find_stack_output(make_required_key_for_ecs_application_url(env), value_only=True)
+    if found:
+        return found
+    else:  # Couldn't this just raise an error? -kmp 17-Aug-2021
+        PRINT('Did not locate the server from Cloudformation! Check ECS Stack metadata.')
+        return ''
```
[the remark this answers is not in the document]
Comment (netsettler) [reply dcicutils/cloudformation_utils.py:42]: Will worry about this in a future PR.

[the remark this answers is not in the document]
Comment (netsettler) [reply dcicutils/cloudformation_utils.py:47]: I added a clear statement in the doc string not to expect good behavior so we don't have to bump the major version if we change it.

**Pull request merged**